programs/pluto/kernel.c

```diff
@@ -3655,3 +3723,97 @@ void shutdown_kernel(struct logger *logger)
+{
+	struct child_sa *child = find_v2_child_sa_by_spi(spi, protoid, dst);
+	address_buf a;
+	const struct connection *c;
+
+	if (child == NULL) {
+		llog(LOG_STREAM, logger, "Received kernel %s EXPIRE event for IPsec SPI 0x%x, but there is no connection with this SPI and dst %s bytes %" PRIu64 " packets %" PRIu64,
+		     hard ? "hard" : "soft",
+		     ntohl(spi), str_address(dst, &a), bytes, packets);
+		return;
+	}
+
+	c = child->sa.st_connection;
+
+	if ((hard && impair.ignore_hard_expire) ||
+	     (!hard && impair.ignore_soft_expire)) {
+                        llog(RC_LOG, c->logger, "IMPAIR is supressing a %s EXPIRE event spi 0x%x dst %s bytes %" PRIu64 " packets %" PRIu64,
+			     hard ? "hard" : "soft", ntohl(spi), str_address(dst, &a),
+			     bytes, packets);
+                        return;
+	}
+
+	bool rekey = !LIN(POLICY_DONT_REKEY, c->policy);
+	bool newest = c->newest_ipsec_sa == child->sa.st_serialno;
+	struct state *st =  &child->sa;
+	struct ipsec_proto_info *pr = st->st_esp.present ? &st->st_esp : st->st_ah.present ? &st->st_ah : st->st_ipcomp.present ? &st->st_ipcomp : NULL;
+
+	bool softexpired = ((pr->peer_kernel_sa_expired & SA_SOFT_EXPIRED) ||
+			(pr->our_kernel_sa_expired & SA_SOFT_EXPIRED));
+
+	bool hardexpired = ((pr->peer_kernel_sa_expired & SA_HARD_EXPIRED) ||
+			(pr->our_kernel_sa_expired & SA_HARD_EXPIRED));
+
+	enum sa_expire_kind expire = hard ? SA_HARD_EXPIRED : SA_SOFT_EXPIRED;
+
+	llog(RC_LOG, c->logger, "Received %s EXPIRE for SPI 0x%x bytes %" PRIu64 " packets %" PRIu64 " %s rekey=%s%s%s%s",
+	     hard ? "hard" : "soft", ntohl(spi), bytes, packets,
+             newest ? "for the newest SA" : "for old SA. Delete it or it is about to expire.",
+	     rekey ?  "yes" : "no",
+	     softexpired ? " one of SA was soft expired ignore this expire." : ".",
+	     hardexpired ? " one of SA was hard expired ignore this expire." : ".",
+             (newest && rekey && !softexpired && !hardexpired) ? " Replace this SA" : "");
+
+	if ((softexpired && expire == SA_SOFT_EXPIRED)  ||
+	    (hardexpired && expire == SA_HARD_EXPIRED)) {
+		dbg("#%lu one of the SA has already expired ignore this %s EXPIRE",
+		    child->sa.st_serialno, hard ? "hard" : "soft");
+		/*
+		 * likely the other direction SA EXPIRED, it triggered first.
+		 * It should be safe to ignore. No need for another log.
+		 */
+	} else if (!hardexpired && expire == SA_HARD_EXPIRED) {
+		if (pr->attrs.spi == spi) {
+			pr->our_kernel_sa_expired |= expire;
+			set_sa_info(pr, bytes, add_time, true /* inbound */, NULL);
+		} else {
+			pr->peer_kernel_sa_expired |= expire;
+			set_sa_info(pr, bytes, add_time, false /* outbound */, NULL);
+		}
+		set_sa_expire_next_event(EVENT_SA_EXPIRE, &child->sa);
+	} else if (newest && rekey && !hardexpired && !softexpired && expire == SA_SOFT_EXPIRED) {
+		if (pr->attrs.spi == spi) {
+			set_sa_info(pr, bytes, add_time, true /* inbound */, NULL);
```

Comment:
I don't how it will set to 0! The expire message contains the bytes counter.
The idea here is to update st counter with bytes counter that came in the kernel expire message.